cdap-app-templates/cdap-etl/hydrator-test/src/main/java/io/cdap/cdap/etl/mock/batch/MockPullProducer.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.etl.mock.batch;
+
+import com.google.common.reflect.TypeToken;
+import com.google.gson.Gson;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.api.spark.sql.DataFrames;
+import io.cdap.cdap.etl.api.Field;
+import io.cdap.cdap.etl.api.engine.sql.dataset.RecordCollection;
+import io.cdap.cdap.etl.api.engine.sql.dataset.SQLDataset;
+import io.cdap.cdap.etl.api.engine.sql.dataset.SQLDatasetDescription;
+import io.cdap.cdap.etl.api.engine.sql.dataset.SQLDatasetProducer;
+import io.cdap.cdap.etl.api.engine.sql.request.SQLPullRequest;
+import io.cdap.cdap.etl.api.sql.engine.dataset.SparkRecordCollectionImpl;
+import org.apache.spark.SparkContext;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SQLContext;
+import org.apache.spark.sql.types.StructType;
+
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Pull Dataset implementation for unit test
+ */
+public class MockPullProducer implements SQLDatasetProducer<StructuredRecord> {
+  private static final Gson GSON = new Gson();
+
+  private final SQLDatasetDescription datasetDescription;
+  private final String expected;
+
+  public MockPullProducer(SQLPullRequest pullRequest, String expected) {
+    this.expected = expected;
+    this.datasetDescription = new SQLDatasetDescription() {
+      @Override
+      public String getDatasetName() {
+        return pullRequest.getDatasetName();
+      }
+
+      @Override
+      public Schema getSchema() {
+        return pullRequest.getDatasetSchema();
+      }
+    };
+  }
+
+  @Override
+  public SQLDatasetDescription getDescription() {
+    return this.datasetDescription;
+  }
+
+  @Override
+  public RecordCollection<StructuredRecord> produce(SQLDataset dataset) {
+    // Create a spark session and write RDD as JSON
+    TypeToken<HashSet<StructuredRecord>> typeToken = new TypeToken<HashSet<StructuredRecord>>() { };
+    Type setOfStructuredRecordType = typeToken.getType();
+
+    // Read records from JSON and adjust data types
+    Set<StructuredRecord> jsonRecords = GSON.fromJson(expected, setOfStructuredRecordType);
```

Comment:
Do you need to go through StructuredRecord here?

Reply:
for the purpose of the test, yes. Unfortunately, GSON deserializes integers as Doubles, so I need to "rebuild" the structured record into the correct data types.

Reply:
Got you. Please make a comment about it